### PR TITLE
fix(native): Enforce NOT NULL column constraints for table writes

### PR DESCRIPTION
**Pull request by Joe-Abraham**
## Description

A native worker ignored `NOT NULL` column constraints on writes: an `INSERT` of `NULL` into a `NOT NULL` column succeeded, while the same query failed on a Java worker.

The plan converter always built `core::InsertTableHandle` with an empty not-null column set, so Velox had nothing to check. This fills it in for `TableWriterNode` and `CallDistributedProcedureNode`.

The plan names the constrained columns by source variable, but Velox matches on target column names. `columns` and `columnNames` are parallel — `columns[i]` is written to `columnNames[i]` — so the converter walks that pairing to map each not-null variable to its target name(s). A not-null variable missing from `columns` is an inconsistent plan and fails as a system error rather than silently dropping the constraint.

## Motivation and Context

Part of #28156, item 2: the native worker now enforces `NOT NULL` on writes. Fixes explicit `NULL`s, `NULL`s from a source column, and `NOT NULL` columns omitted from the insert column list. The converter is connector-agnostic, so Iceberg is covered too; Iceberg `DROP NOT NULL` and its native tests follow in a separate PR.

## Impact

Inserts that used to write `NULL` into a `NOT NULL` column now fail, as they already did on Java. No configuration; enforcement follows table metadata.

The message contains the Java text `NULL value not allowed for NOT NULL column: <column>`, but Velox uses `VELOX_USER_CHECK`, so the translator wraps it:

```
!notNullDecodedVector_.hasNulls() NULL value not allowed for NOT NULL column: c2 Operator: TableWrite[...]
```

The code differs too — `GENERIC_USER_ERROR` instead of Java's `CONSTRAINT_VIOLATION`. Both are `USER_ERROR`, so failure classification is unchanged.

## Test Plan

`TestPrestoNativeWriter` — write path, the violation asserted on native and Java:

* `testNotNullColumnEnforcement` — explicit `NULL`, omitted column, `NULL`s from a source column, reordered insert column list; also that valid rows still write and a rejected insert writes nothing.
* `testNotNullColumnEnforcementWithSharedSourceVariable` — one source column into two `NOT NULL` targets.
* `testCreateTableAsSelectDoesNotInheritNotNull` — `CREATE TABLE AS` has no column constraint syntax and does not carry the source table's constraints over, so the new table is unconstrained.
* `testAlterColumnSetAndDropNotNull` — `ALTER COLUMN ... SET/DROP NOT NULL` taking effect on the next insert.

`TestRewriteDataFilesProcedure.testRewriteDataFilesOnTableWithNotNullColumn` covers the `CallDistributedProcedureNode` call site, where the constrained set is now non-empty for the first time.

The native worker names the column by target column, while the java worker names it by the source channel of its plan variable, so only the native assertion pins the column name.

`toNotNullColumnNames` is file-local to `PrestoToVeloxQueryPlan.cpp` and is covered through the write path rather than by unit tests. Its two `VELOX_CHECK`s — a `NOT NULL` variable missing from the TableWriter source variables, and a `columns`/`columnNames` size mismatch — guard inconsistent coordinator plans that no SQL statement can produce, so they are deliberately left untested.

`PrestoNativeQueryRunnerUtils` sets `hive.allow-add-constraint`, which the native runner needs for that `ALTER TABLE` under `legacy` security. It is bound only under legacy security, so it stays off the java runner.

```
mvn test -pl presto-native-execution -Dtest='TestPrestoNativeWriter#testNotNullColumnEnforcement+testNotNullColumnEnforcementWithSharedSourceVariable+testCreateTableAsSelectDoesNotInheritNotNull+testAlterColumnSetAndDropNotNull'

mvn test -pl presto-native-execution -Dtest='TestRewriteDataFilesProcedure#testRewriteDataFilesOnTableWithNotNullColumn'
```

## Contributor checklist

- [x] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.
- [x] If adding new dependencies, verified they have an [OpenSSF Scorecard](https://securityscorecards.dev/#the-checks) score of 5.0 or higher (or obtained explicit TSC approval for lower scores).

## Release Notes

```
== RELEASE NOTES ==

Prestissimo (Native Execution) Changes
* Fix ``NOT NULL`` column constraints being ignored for table writes, which allowed an ``INSERT`` to silently write ``NULL`` into a ``NOT NULL`` column.
```

## Summary by Sourcery

Enforce table-level NOT NULL constraints consistently across native table-writing paths.

Bug Fixes:
- Enforce NOT NULL constraints for native table writes, preventing INSERT operations from silently writing NULL values to constrained columns.
- Apply NOT NULL enforcement to distributed procedure table writes as well as regular table writes.

Enhancements:
- Map constrained source variables to their target table columns, including reordered insert columns and shared source variables, while rejecting inconsistent plans.

Tests:
- Add native and Java worker coverage for explicit NULLs, omitted columns, source-derived NULLs, reordered inserts, shared variables, valid writes, and rejected-write atomicity.
- Verify constraint changes through ALTER TABLE, CTAS constraint behavior, and Iceberg data-file rewrite procedures.

### CODE REVIEW

**Comment by Joe-Abraham** [comment]
@sourcery-ai review

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 1]
```diff
@@ -1567,6 +1567,52 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+  for (const auto& variable : notNullColumnVariables) {
+    notNullVariableNames.insert(variable.name);
+  }
+
+  // A source variable can feed more than one target column, so translate by
+  // position: every column whose source variable is constrained contributes its
+  // own target name.
+  folly::F14FastSet<std::string> translatedVariableNames;
+  translatedVariableNames.reserve(notNullVariableNames.size());
+  notNullColumnNames.reserve(columns.size());
```
Comment (nmahadevuni) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp]: Can the size of notNullColumnNames be more than notNullVariableNames? If not, do `min(notNullVariableNames.size(), columns.size())`

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeWriter.java [hunk 2]
```diff
@@ -280,6 +280,85 @@ public void testInsertIntoSpecialPartitionName()
         }
     }
 
+    @Test(groups = {"writer"})
+    public void testNotNullColumnEnforcement()
```
Comment (nmahadevuni) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeWriter.java:284]: Please add a test for CREATE TABLE AS too.

**Review by nmahadevuni — COMMENTED** [review]
Thanks @Joe-Abraham. Had a couple comments.

File: presto-native-execution/presto_cpp/main/types/tests/PrestoToVeloxQueryPlanTest.cpp [hunk 3]
```diff
@@ -44,6 +45,13 @@ void validateSqlFunctionHandleParsing(
     EXPECT_EQ(*expectedRawInputTypes[i], *actualRawInputTypes[i]);
   }
 }
+
+protocol::VariableReferenceExpression makeVariable(const std::string& name) {
+  protocol::VariableReferenceExpression variable;
+  variable.name = name;
+  variable.type = "bigint";
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/tests/PrestoToVeloxQueryPlanTest.cpp]: Make this a function parameter as well instead of hardcoding to bigint.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h [hunk 4]
```diff
@@ -354,4 +356,18 @@ void parseIndexLookupCondition(
     bool acceptConstant,
     std::vector<velox::core::IndexLookupConditionPtr>& joinConditionPtrs,
     std::vector<velox::core::TypedExprPtr>& unsupportedConditions);
+
+/// Translates 'notNullColumnVariables' (source variable names) to the target
+/// column names that core::InsertTableHandle::notNullColumns() and the
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h]: Maybe use a 'source' and 'targetTable' based naming instead of columns, columnNames to make the code more readable.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 5]
```diff
@@ -1567,6 +1567,54 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+  for (const auto& variable : notNullSourceVariables) {
+    notNullSourceVariableNames.insert(variable.name);
+  }
+
+  // A source variable can feed more than one target table column, so translate
+  // by position: every constrained source variable contributes the name of each
+  // target table column it is written to.
+  folly::F14FastSet<std::string> translatedSourceVariableNames;
+  translatedSourceVariableNames.reserve(notNullSourceVariableNames.size());
+  notNullTargetTableColumnNames.reserve(sourceVariables.size());
```
[the remark this answers is not in the document]
Comment (Joe-Abraham) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:1597]: @nmahadevuni, should we revert the code?

**Comment by Joe-Abraham** [comment]
@aditi-pandit, Thank you for your feedback. I’ve addressed your comments. Please take a look when you have some time. Thanks.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h [hunk 6]
```diff
@@ -354,4 +356,20 @@ void parseIndexLookupCondition(
     bool acceptConstant,
     std::vector<velox::core::IndexLookupConditionPtr>& joinConditionPtrs,
     std::vector<velox::core::TypedExprPtr>& unsupportedConditions);
+
+/// Translates 'notNullSourceVariables' to the target table column names that
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h]: This doesn't quite explain what the returned value is.

Also "A source variable may feed several target table
/// columns, in which case all of those column names are returned. " is confusing.
The one-one mapping expectation from the previous paragraph implies that a source Variable reference should be repeated as well. So if that repeated sourceVariable is not null then what are the implications ?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 7]
```diff
@@ -1567,6 +1567,54 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+      notNullTargetTableColumnNames.insert(targetTableColumnNames[i]);
+      translatedSourceVariableNames.insert(sourceVariables[i].name);
+    }
+  }
+
+  // The planner derives notNullColumnVariables from the same list as the
+  // TableWriter source variables, so a miss is an inconsistent plan from the
+  // coordinator rather than bad user input. Fail loudly instead of dropping the
+  // constraint silently.
+  for (const auto& name : notNullSourceVariableNames) {
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp]: It might be better to add this check in the first loop itself that is iterating over the notNullSourceVariables to extract the names.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h [hunk 8]
```diff
@@ -354,4 +356,20 @@ void parseIndexLookupCondition(
...
+/// core::InsertTableHandle::notNullColumns() and the TableWriter operator match
+/// against. 'sourceVariables' and 'targetTableColumnNames' are parallel: the
+/// value of 'sourceVariables[i]' is written to target table column
+/// 'targetTableColumnNames[i]'. A source variable may feed several target table
+/// columns, in which case all of those column names are returned. Raises a
+/// system error if a not-null source variable is absent from 'sourceVariables',
+/// since enforcement for that column would otherwise be dropped. Exposed for
+/// testing.
+folly::F14FastSet<std::string> toNotNullColumnNames(
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h]: Did you make this a public function so that it can be used directly in tests ? We should avoid that.

I'm actually fine if you skip the PrestoToVeloxQueryPlanTest unit tests or make them VELOX_FRIEND_TEST.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeWriter.java [hunk 9]
```diff
@@ -280,6 +280,109 @@ public void testInsertIntoSpecialPartitionName()
...
+        String tmpTableName = generateRandomTableName();
+        try {
+            getQueryRunner().execute(format("CREATE TABLE %s (c1 BIGINT, c2 BIGINT NOT NULL)", sourceTableName));
+            assertUpdate(format("INSERT INTO %s VALUES (NULL, 2)", sourceTableName), 1);
+
+            // CREATE TABLE AS has no column constraint syntax and does not carry the source table's
+            // constraints over, so every column of the new table is nullable and the write is
+            // unconstrained even where it reads a NOT NULL column.
+            assertUpdate(format("CREATE TABLE %s AS SELECT c1, c2 FROM %s", tmpTableName, sourceTableName), 1);
+            assertQuery(format("SELECT c1, c2 FROM %s", tmpTableName), "VALUES (CAST(NULL AS BIGINT), CAST(2 AS BIGINT))");
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeWriter.java:349]: Does this constraint enforcement also happen for MERGE SQL ? Can you try that.

Comment (Joe-Abraham) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeWriter.java:349]: Tried it — MERGE does enforce the constraint, just not through the path this PR touches.

`MergeWriterNode` doesn't carry `notNullColumnVariables` at all. The coordinator instead wraps each non-nullable column of the merge row in `COALESCE(expr, CAST(fail('NULL value not allowed for NOT NULL column...') AS type))` (see `QueryPlanner.checkNotNullColumns`), for both the `WHEN MATCHED ... UPDATE SET` and `WHEN NOT MATCHED ... INSERT` branches. That's just a projection, so the native worker evaluates it like any other expression.

Ran it against a local native cluster on Iceberg:

- `WHEN MATCHED THEN UPDATE SET val = s.val` where `s.val` is NULL -> rejected with `NULL value not allowed for NOT NULL column. Table: merge_target Column: val`
- `WHEN NOT MATCHED THEN INSERT (id, val) VALUES (s.id, s.val)` with NULL -> same rejection
- same MERGE with a non-null value -> succeeds, row written

One issue I encountered was that MERGE requires #28468. If it’s missing, the plan conversion fails on the worker before it even reaches the check.

So this is already covered and unaffected by the change here. The Java side has `IcebergDistributedTestBase.testMergeNonNullableColumns` for the same cases. I can add a native equivalent if you'd like, but it would be the first native MERGE e2e test in the repo, so it might sit better as a follow-up with the rest of the native UPDATE/MERGE coverage.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/AbstractTestRewriteDataFilesProcedure.java [hunk 10]
```diff
@@ -28,6 +28,34 @@
...
+            // create 3 files
+            assertUpdate("INSERT INTO " + tableName + " values(1, 'foo'), (2, 'bar')", 2);
+            assertUpdate("INSERT INTO " + tableName + " values(3, 'foo'), (4, 'bar')", 2);
+            assertUpdate("INSERT INTO " + tableName + " values(5, 'foo'), (6, 'bar')", 2);
+            validateDataFilesAndDeleteFiles(tableName, 3L, 0L);
+
+            // The procedure writes through a CallDistributedProcedureNode, which carries the
+            // table's NOT NULL columns like any other write. Rows already satisfy the constraint,
+            // so rewriting them keeps every row.
+            assertUpdate(format("CALL system.rewrite_data_files(table_name => '%s', schema => '%s', options => map(array['rewrite-all'], array['true']))", tableName, schemaName), 6);
```
[the remark this answers is not in the document]
Comment (Joe-Abraham) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/iceberg/AbstractTestRewriteDataFilesProcedure.java:48]: Tried this, but it is not reachable on Iceberg: `ALTER COLUMN ... SET NOT NULL` fails with `Catalog iceberg does not support ALTER COLUMN with NOT NULL`, and `ADD COLUMN ... NOT NULL` is rejected unless it has a DEFAULT, so there is no way to get pre-existing NULLs into a NOT NULL column. Worth noting `SET NOT NULL` is metadata-only anyway (even on Hive, where it works) - it never validates existing rows, only later writes fail.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 11]
```diff
@@ -1663,7 +1732,10 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
   }
 
   auto insertTableHandle = std::make_shared<core::InsertTableHandle>(
-      connectorId, connectorInsertHandle, folly::F14FastSet<std::string>{});
+      connectorId,
+      connectorInsertHandle,
+      toNotNullColumnNames(
+          node->notNullColumnVariables, node->columns, node->columnNames));
```
[the remark this answers is not in the document]
Comment (Joe-Abraham) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:1738]: I do not think this can happen: Velox matches `notNullColumns` against `tableWriteNode->columnNames()`, which is the same `node->columnNames` list we build them from, so the two stay in sync whatever SymbolMapper put in it.

**Review by aditi-pandit — COMMENTED** [review]
This PR looks good. 

@mohsaka recently added support to ALTER column order for schema evolution https://github.com/prestodb/presto/pull/28349. Can you also add some test cases to ensure that the NOT NULL constraints are checked correctly after changing the column order.

**Comment by Joe-Abraham** [comment]
> This PR looks good.
> 
> @mohsaka recently added support to ALTER column order for schema evolution #28349. Can you also add some test cases to ensure that the NOT NULL constraints are checked correctly after changing the column order.

@aditi-pandit, Could I include that within https://github.com/prestodb/presto/pull/28040 alongside the full alter column set not null changes and ship this seperate?

**Review by aditi-pandit — APPROVED** [review]
That is fine. Thanks @Joe-Abraham
